forc-plugins/forc-client/src/util/aws.rs

```diff
@@ -0,0 +1,269 @@
+}
+
+async fn request_get_pubkey(
+    kms: &Client,
+    key_id: String,
+) -> std::result::Result<GetPublicKeyOutput, anyhow::Error> {
+    kms.get_public_key()
+        .key_id(key_id)
+        .send()
+        .await
+        .map_err(Into::into)
+}
+
+/// Decode an AWS KMS Pubkey response.
+fn decode_pubkey(resp: &GetPublicKeyOutput) -> std::result::Result<Vec<u8>, anyhow::Error> {
+    let raw = resp
+        .public_key
+        .as_ref()
+        .ok_or(anyhow::anyhow!("public key not found"))?;
+    Ok(raw.clone().into_inner())
+}
+
+async fn sign_with_kms(
+    client: &aws_sdk_kms::Client,
+    key_id: &str,
+    public_key_bytes: &[u8],
+    message: Message,
+) -> anyhow::Result<fuel_crypto::Signature> {
+    use k256::{
+        ecdsa::{RecoveryId, VerifyingKey},
+        pkcs8::DecodePublicKey,
+    };
+
+    let reply = client
+        .sign()
+        .key_id(key_id)
+        .signing_algorithm(SigningAlgorithmSpec::EcdsaSha256)
+        .message_type(MessageType::Digest)
+        .message(Blob::new(*message))
+        .send()
+        .await
+        .inspect_err(|err| tracing::error!("Failed to sign with AWS KMS: {err:?}"))?;
+    let signature_der = reply
+        .signature
+        .ok_or_else(|| anyhow::anyhow!("no signature returned from AWS KMS"))?
+        .into_inner();
+    // https://stackoverflow.com/a/71475108
+    let sig = k256::ecdsa::Signature::from_der(&signature_der)
+        .map_err(|_| anyhow::anyhow!("invalid DER signature from AWS KMS"))?;
+    let sig = sig.normalize_s().unwrap_or(sig);
+
+    // This is a hack to get the recovery id. The signature should be normalized
+    // before computing the recovery id, but aws kms doesn't support this, and
+    // instead always computes the recovery id from non-normalized signature.
+    // So instead the recovery id is determined by checking which variant matches
+    // the original public key.
+
+    let recid1 = RecoveryId::new(false, false);
+    let recid2 = RecoveryId::new(true, false);
+
+    let rec1 = VerifyingKey::recover_from_prehash(&*message, &sig, recid1);
+    let rec2 = VerifyingKey::recover_from_prehash(&*message, &sig, recid2);
+
+    let correct_public_key = k256::PublicKey::from_public_key_der(public_key_bytes)
+        .map_err(|_| anyhow::anyhow!("invalid DER public key from AWS KMS"))?
+        .into();
+
+    let recovery_id = if rec1.map(|r| r == correct_public_key).unwrap_or(false) {
+        recid1
+    } else if rec2.map(|r| r == correct_public_key).unwrap_or(false) {
+        recid2
+    } else {
+        anyhow::bail!("Invalid signature generated (reduced-x form coordinate)");
+    };
+
+    // Insert the recovery id into the signature
+    debug_assert!(
+        !recovery_id.is_x_reduced(),
+        "reduced-x form coordinates are caught by the if-else chain above"
+    );
+    let v = recovery_id.is_y_odd() as u8;
+    let mut signature = <[u8; 64]>::from(sig.to_bytes());
+    signature[32] = (v << 7) | (signature[32] & 0x7f);
```

Comment:
What is this doing? It would be great to have more detailed comments here.

Reply:
agree some comments here would be very beneficial